File: Task/Extract-file-extension/Visual-Basic/extract-file-extension.vb
Option Explicit
'-----------------------------------------------------------------
Function ExtractFileExtension(ByVal Filename As String) As String
Dim i As Long
Dim s As String
  i = InStrRev(Filename, ".")
  If i Then
    If i < Len(Filename) Then
      s = Mid$(Filename, i)
      For i = 2 To Len(s)
        Select Case Mid$(s, i, 1)
        Case "A" To "Z", "a" To "z", "0" To "9"
            'these characters are OK in an extension; continue
        Case Else
            'this one is not OK in an extension
            Exit Function
        End Select
      Next i
      ExtractFileExtension = s
    End If
  End If
End Function
'-----------------------------------------------------------------
Sub Main()
Dim s As String
  s = "http://example.com/download.tar.gz"
  Debug.Assert ExtractFileExtension(s) = ".gz"
  s = "CharacterModel.3DS"
  Debug.Assert ExtractFileExtension(s) = ".3DS"
  s = ".desktop"
  Debug.Assert ExtractFileExtension(s) = ".desktop"
  s = "document"
  Debug.Assert ExtractFileExtension(s) = ""
  s = "document.txt_backup"
  Debug.Assert ExtractFileExtension(s) = ""
  s = "/etc/pam.d/login"
  Debug.Assert ExtractFileExtension(s) = ""
  s = "desktop."
  Debug.Assert ExtractFileExtension(s) = ""
  s = "a.~.c"
  Debug.Assert ExtractFileExtension(s) = ".c"
  s = "a.b.~"
  Debug.Assert ExtractFileExtension(s) = ""
  s = "a.b.1~2"
  Debug.Assert ExtractFileExtension(s) = ""
End Sub

File: Task/Extract-file-extension/Free-Pascal-Lazarus/extract-file-extension.pas
Program Extract_file_extension;

{FreePascal has the built-in function ExtractFileExt which returns the file extension.
* it does need charachters before the period to return the proper extension and it returns
* the extension including the period}

Uses character,sysutils;

Const arr : array of string = ('http://example.com/download.tar.gz','CharacterModel.3DS','.desktop',
                               'document','document.txt_backup','/etc/pam.d/login');

Function extractextension(fn: String): string;
Var
  i: integer;
Begin
  fn := 'prefix' + fn; {add charachters before the period}
  fn := ExtractFileExt(fn);
  For i := 2 to length(fn) Do {skip the period}
    If Not IsLetterOrDigit(fn[i]) Then exit('');
 extractextension := fn;
End;

Var i : string;
Begin
  For i In arr Do
    writeln(i:35,' -> ',extractextension(i))
End.

